lib/pbench/test/unit/server/query_apis/test_datasets_publish.py

```diff
@@ -83,16 +84,32 @@ class TestDatasetsPublish(Commons):
     In a web service context, we access class functions mostly via the
     Flask test client rather than trying to directly invoke the class
     constructor and `post` service.
+
+    FIXME:
+    We need to re-initialize the database session with expire_on_commit parameter set to false.
+    Otherwise we get sqlalchemy.orm.exc.DetachedInstanceError: Instance <User> is not bound to a Session;
+    attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3).
+    The reason for this might involve with the fact that we are accessing 2 different users using the
+    token (admin token) generated by another fixture that uses different database session.
+
+    In our test_query function the database session might expire after successful query (with admin token)
+    but 2 user instances might get left behind without getting properly expunged.
+
+    This often happens due to objects being in expired state, objects get expired after committing,
+    then when such expired objects are about to get used the ORM tries to refresh them, but this cannot be done
+    when objects are detached from session (e.g. because that session was closed). This behavior can be managed
+    by creating session with expire_on_commit=False param.
```

Comment:
Ugh.

I don't like twisting the production code `init_db` to deal with issues like this. There's got to be a better way, probably including a more thorough cleanup of the database. For example, using `yield` in the DB setup fixtures so that we can completely clean up after each test cycle.

And, again, if we can move towards a pure unit test environment that doesn't involve an actual DB, we wouldn't have to deal with this sort of issue at all ...

Reply:
Obviously, for unit testing, mocking out the database methods would be the appropriate and superior approach.

Short of that, it seems like what is required is proper management of the database session -- i.e., create it at the beginning of the test, maintain it through the test, and the close it down when we're done (using a fixture with a `yield` like Dave suggests seems an excellent way to implement this).  But, yeah, it means that the production code has to be prepared for persistent sessions rather than relying on ephemeral ones.

Reply:
Yeah its ugly unfortunately. However, I fixed this problem of expired orm instances by expunging before everytime we add a new user. Sqlalchemy docs says that a detached instance is an instance which corresponds, or previously corresponded, to a record in the database, but is not currently in any session and so we need to expunge them before adding.

I would also like to mention that the none of the test actually fails because of this  `DetachedInstanceError` it just prints the warning with big call stack and we probably garbage collect all the instances after the tests are finished but the warning was still annoying. Also It will take massive amount of rework in our testing architecture to get rid of database calls that we have at the moment, probably a task for another sprint.